### PR TITLE
[AutoDiff upstream] AST bits for @differentiable fn ty

### PR DESCRIPTION
**Pull request by marcrasi**
Adds the AST bits necessary to represent `@differentiable` `AnyFunctionType` and `SILFunctionType`. Plumbs these bits through parsing and printing so that they can be tested. Adds serialization for them.

This resolves TF-871.

Intentionally omits any functionality based on these bits to keep this PR small. (e.g. typechecking that the `@differentiable` attr is allowed on the types; lowering `@differentiable` `AnyFunctionType` to the corresponding `@differentiable` `SILFunctionType`)

### CODE REVIEW

File: include/swift/AST/Types.h [hunk 1]
```diff
@@ -3750,6 +3774,16 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
     // Is this function guaranteed to be no-escape by the type system?
     bool isNoEscape() const { return Bits & NoEscapeMask; }
 
+    bool isDifferentiable() const {
+      return getDifferentiabilityKind() >
```
Comment (lattner) [inline include/swift/AST/Types.h]: nit, but it seems like != would be more natural here than >

File: lib/AST/ASTDemangler.cpp [hunk 2]
```diff
@@ -479,9 +479,9 @@ Type ASTBuilder::createImplFunctionType(
     break;
   }
 
-  auto einfo = SILFunctionType::ExtInfo(representation,
-                                        flags.isPseudogeneric(),
-                                        !flags.isEscaping());
+  auto einfo = SILFunctionType::ExtInfo(
```
Comment (lattner) [inline lib/AST/ASTDemangler.cpp]: Is there a mangling for these flags?  If so, it seems like it should be plumbed in here for the demangler to use.

**Review by lattner — APPROVED** [review]
LG thanks!

File: lib/AST/ASTDemangler.cpp (hunk 2 above)
Comment (marcrasi) [reply lib/AST/ASTDemangler.cpp]: We've got some mangling in the `tensorflow` branch. I will upstream it in a separate PR. I'll add a TODO comment here referencing the ticket: https://bugs.swift.org/browse/TF-948

File: include/swift/AST/Types.h (hunk 1 above)
Comment (marcrasi) [reply include/swift/AST/Types.h]: Fixed

**Comment by marcrasi** [comment]
@swift-ci please test and merge

**Comment by marcrasi** [comment]
@swift-ci please test and merge

**Comment by marcrasi** [comment]
@swift-ci please test and merge

**Comment by marcrasi** [comment]
@swift-ci please test and merge

**Comment by marcrasi** [comment]
@swift-ci please test and merge

**Comment by marcrasi** [comment]
@swift-ci please test and merge

**Pull request merged**